returns/interfaces/specific/failable.py

```diff
@@ -0,0 +1,165 @@
+from returns.interfaces.container import ContainerN
+from returns.interfaces.rescuable import RescuableN
+from returns.primitives.asserts import assert_equal
+from returns.primitives.hkt import KindN
+from returns.primitives.laws import (
+    Law,
+    Law3,
+    Lawful,
+    LawSpecDef,
+    law_definition,
+)
+
+_FirstType = TypeVar('_FirstType')
+_SecondType = TypeVar('_SecondType')
+_ThirdType = TypeVar('_ThirdType')
+
+_DiverseFailableType = TypeVar('_DiverseFailableType', bound='DiverseFailableN')
+
+# Used in laws:
+_NewFirstType = TypeVar('_NewFirstType')
+
+
+@final
+class _FailableLawSpec(LawSpecDef):
+    """
+    Failable laws.
+
+    We need to be sure that ``.rescue`` won't rescue success types.
+    """
+
+    @law_definition
+    def rescue_short_circuit_law(
+        raw_value: _FirstType,
+        container: 'FailableN[_FirstType, _SecondType, _ThirdType]',
+        function: Callable[
+            [_SecondType],
+            KindN['FailableN', _FirstType, _NewFirstType, _ThirdType],
+        ],
+    ) -> None:
+        """Ensures that you cannot rescue a success."""
+        assert_equal(
+            container.from_value(raw_value),
+            container.from_value(raw_value).rescue(function),
+        )
+
+
+@final
+class _DiverseFailableLawSpec(LawSpecDef):
+    """
+    Diverse Failable laws.
+
+    We need to be sure that ``.map``, ``.bind`` and ``.alt``
+    works correctly for both success and failure types.
+    """
+
+    @law_definition
+    def map_short_circuit_law(
+        raw_value: _SecondType,
+        container: 'DiverseFailableN[_FirstType, _SecondType, _ThirdType]',
+        function: Callable[[_FirstType], _NewFirstType],
+    ) -> None:
+        """Ensures that you cannot map a failure."""
+        assert_equal(
+            container.from_failure(raw_value),
+            container.from_failure(raw_value).map(function),
+        )
+
+    @law_definition
+    def bind_short_circuit_law(
+        raw_value: _SecondType,
+        container: 'DiverseFailableN[_FirstType, _SecondType, _ThirdType]',
+        function: Callable[
+            [_FirstType],
+            KindN['DiverseFailableN', _NewFirstType, _SecondType, _ThirdType],
+        ],
+    ) -> None:
+        """
+        Ensures that you cannot bind a failure.
+
+        See: https://wiki.haskell.org/Typeclassopedia#MonadFail
+        """
+        assert_equal(
+            container.from_failure(raw_value),
+            container.from_failure(raw_value).bind(function),
+        )
+
+    @law_definition
+    def alt_short_circuit_law(
+        raw_value: _SecondType,
+        container: 'DiverseFailableN[_FirstType, _SecondType, _ThirdType]',
+        function: Callable[[_SecondType], _NewFirstType],
+    ) -> None:
+        """Ensures that you cannot alt a success."""
+        assert_equal(
+            container.from_value(raw_value),
+            container.from_value(raw_value).alt(function),
+        )
+
+
+class FailableN(
```

Comment:
One more thing. I really dislike current `Maybe` situation. It has no way to directly access a failure case from its class. And we have to import `Nothing`.

`DiverseFailable` has `.from_failure`, `Failable` should have `empty` property instead. So, we would be able to add `empty: ClassVar[Maybe[Any]] = Nothing` to `Maybe`!

Reply:
Is this worth it?? Because `empty` is `Maybe` specific behavior, if I have a class like:
```python
class Number(Failable[_ValueType, float('nan')]):
    ...

n: Number[10.0]
print(n.empty) # => nan
```

Does not make much sense to me!

Reply:
Well, this is kind of symmetric to what we have in `Result`. And `empty` can later be used with `monoid` if we would ever need one.

Reply:
Seems weird when we think about `Result`, like, what should `Result.empty` means?
Can you provide a practical example without using `Maybe`??

Reply:
Oh, I see. `Failable` would require `Result` to also have this prop. 
Let's discard this idea for now.